tests/python/pants_test/rules/test_test.py

```diff
@@ -102,17 +97,7 @@ def test_coordinator_python_test(self):
     target_adaptor = PythonTestsAdaptor(type_alias='python_tests')
 
     result = run_rule(coordinator_of_tests, HydratedTarget(Address.parse("some/target"), target_adaptor, ()), {
-      (PyTestResult, HydratedTarget): lambda _: PyTestResult(status=Status.FAILURE, stdout='foo'),
+      (TestResult, PythonTestsAdaptor): lambda _: TestResult(status=Status.FAILURE, stdout='foo'),
     })
 
     self.assertEqual(result, TestResult(status=Status.FAILURE, stdout='foo'))
-
-  def test_coordinator_unknown_test(self):
```

Comment:
Hah. You got me.

It will fail at runtime with a slightly awkward error message (that I attempted to improve over here: #7502)... namely, that there is no declared `Get(TestResult, $type)`.

Reply:
I suspect that if we continue to improve that error message code path by either/or:
1. using more readable `@rule` signatures (as in #7509),
2. allowing `@union`s to be directly included in error messages (which I encouraged @cosmicexplorer not to do in the original union PR, in order to keep things simpler)
... that this can become significantly more friendly.

With 2. in particular (and perhaps with something that consumes the docstring of the union, or takes a description to the union annotation?), something like:
```
Type $x is not a member of the TestTarget @union ($description).
```
... would be possible.

Reply:
I'm not quite sure whether this would be a mostly-rule-author-facing error, or whether it could also leak into general users... If the former, either of those sound good. If the latter, we should be a lot more verbose and explainy...

Either way, this sounds good. Happy to punt on this until some-time-this-week in the interests of getting this merged and avoiding merge conflicts for @Eric-Arellano, but I'd like us to come back to this _soon_ :)